src/pybind/mgr/zabbix/module.py

```diff
@@ -0,0 +1,209 @@
+            data['num_mon'] = len(mon_status['monmap']['mons'])
+        except (KeyError, ValueError):
+            pass
+
+        try:
+            df = self.get('df')
+
+            data['num_pools'] = len(df['pools'])
+            data['total_objects'] = df['stats']['total_objects']
+            data['total_used_bytes'] = df['stats']['total_used_bytes']
+            data['total_bytes'] = df['stats']['total_bytes']
+            data['total_avail_bytes'] = df['stats']['total_avail_bytes']
+
+            wr_ops = 0
+            rd_ops = 0
+            wr_bytes = 0
+            rd_bytes = 0
+
+            for pool in df['pools']:
+                wr_ops += pool['stats']['wr']
+                rd_ops += pool['stats']['rd']
+                wr_bytes += pool['stats']['wr_bytes']
+                rd_bytes += pool['stats']['rd_bytes']
+
+            data['wr_ops'] = wr_ops
+            data['rd_ops'] = rd_ops
+            data['wr_bytes'] = wr_bytes
+            data['rd_bytes'] = rd_bytes
+        except (KeyError, ValueError):
+            pass
+
+        try:
+            osd_map = self.get('osd_map')
+            data['num_osd'] = len(osd_map['osds'])
+
+            num_up = 0
+            num_in = 0
+            for osd in osd_map['osds']:
+                if osd['up'] == 1:
+                    num_up += 1
+
+                if osd['in'] == 1:
+                    num_in += 1
+
+            data['num_osd_up'] = num_up
+            data['num_osd_in'] = num_in
+        except (KeyError, ValueError):
+            pass
+
+        try:
+            pg_summary = self.get('pg_summary')
+            num_pg = 0
+            for state, num in pg_summary['all'].items():
+                num_pg += num
+
+            data['num_pg'] = num_pg
+        except (KeyError, ValueError):
+            pass
+
+        return data
+
+    def shutdown(self):
+        self.log.info('Stopping zabbix')
+        self.run = False
+        self.q.put(None)
+
+    def serve(self):
+        self.log.debug('Zabbix module starting up')
+        self.run = True
+
+        conf = self.get_module_config()
+
+        for key, value in conf.items():
+            self.log.debug('%s: %s', key, value)
+
+        zabbix = ZabbixSender(conf['zabbix_sender'], conf['zabbix_host'],
+                              conf['zabbix_port'], self.log)
+
+        while self.run:
+            self.log.debug('Waking up for new iteration')
+
+            data = self.get_data()
+
+            self.log.debug('Sending data to Zabbix server %s',
+                           conf['zabbix_host'])
+            self.log.debug(data)
+            try:
+                zabbix.send(conf['hostname'], data)
+            except Exception as exc:
+                self.log.error('Exception when sending: %s', exc)
+
+            self.log.debug('Sleeping for {0} seconds'.format(conf['interval']))
+            try:
+                self.q.get(timeout=conf['interval'])
```

Comment:
Yes, it is. I was looking for a non-blocking way to break out of the while True loop. I'll look into threading.Event